hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieHFileReader.java

```diff
@@ -133,21 +138,43 @@ public BloomFilter readBloomFilter() {
 
   @Override
   public Set<String> filterRowKeys(Set candidateRowKeys) {
-    // Current implementation reads all records and filters them. In certain cases, it many be better to:
-    //  1. Scan a limited subset of keys (min/max range of candidateRowKeys)
-    //  2. Lookup keys individually (if the size of candidateRowKeys is much less than the total keys in file)
     try {
-      List<Pair<String, R>> allRecords = readAllRecords();
-      Set<String> rowKeys = new HashSet<>();
-      allRecords.forEach(t -> {
-        if (candidateRowKeys.contains(t.getFirst())) {
-          rowKeys.add(t.getFirst());
-        }
-      });
-      return rowKeys;
+      return filterKeys(new TreeSet<>(candidateRowKeys));
     } catch (IOException e) {
-      throw new HoodieIOException("Failed to read row keys from " + path, e);
+      LOG.error("Failed to fetch keys from " + path);
+    }
+    return Collections.emptySet();
+  }
+
+  @Override
+  public Set<String> filterKeys(TreeSet<String> sortedCandidateRowKeys) throws IOException {
+    return filterRecordsImpl(sortedCandidateRowKeys).keySet();
```

Comment:
There is a faster way to simply filter keys - when scanning to a key, if the key does not exist then the scan fails. So there is no need to read the entire records.

Reply:
Fixed the interface and filterKeys as per the above suggestion